internal/controller/namespacelabel_controller.go

```diff
@@ -0,0 +1,224 @@
+
+		default:
+			r.Log.Info("Adding label", "key", key, "value", value)
+			updatedLabels[key] = value
+		}
+	}
+	return updatedLabels, skippedLabels, duplicateLabels
+}
+
+func (r *NamespacelabelReconciler) updateStatus(ctx context.Context, namespaceLabel *labelsv1.Namespacelabel, updatedLabels, skippedLabels, duplicateLabels map[string]string) error {
+	namespaceLabel.Status.AppliedLabels = updatedLabels
+	namespaceLabel.Status.SkippedLabels = skippedLabels
+	if len(skippedLabels) > 0 {
+		r.SetCondition(namespaceLabel, "LabelsSkipped", metav1.ConditionTrue, "ProtectedLabelsHandled", "Some labels were skipped because they are protected.")
+	} else {
+		r.SetCondition(namespaceLabel, "LabelsSkipped", metav1.ConditionFalse, "ProtectedLabelsHandled", "All labels were applied successfully; no protected labels were skipped.")
+	}
+
+	if len(duplicateLabels) > 0 {
+		r.SetCondition(namespaceLabel, "DuplicateLabels", metav1.ConditionTrue, "DuplicateLabelsHandled", "Some labels were not applied because they are duplicates.")
+	} else {
+		r.SetCondition(namespaceLabel, "DuplicateLabels", metav1.ConditionFalse, "DuplicateLabelsHandled", "All labels were unique and applied successfully.")
+	}
+
+	r.SetCondition(namespaceLabel, "LabelsApplied", metav1.ConditionTrue, "LabelsReconciled", "Labels reconciled successfully.")
+	return r.Status().Update(ctx, namespaceLabel)
+}
+
+func (r *NamespacelabelReconciler) ensureFinalizer(ctx context.Context, namespaceLabel *labelsv1.Namespacelabel) error {
+	if err := finalizer.Ensure(ctx, r.Client, namespaceLabel, r.Log); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (r *NamespacelabelReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&labelsv1.Namespacelabel{}).
+		Watches(&corev1.Namespace{},
+			handler.EnqueueRequestsFromMapFunc(r.enqueueRequestsFromNamespace),
+			builder.WithPredicates(predicate.ResourceVersionChangedPredicate{}),
+		).
+		Complete(r)
+}
+
+// enqueueRequestsFromNamespace triggers reconciliation for related Namespacelabel resources when a Namespace changes.
+// enqueueRequestsFromNamespace reconciles the Namespacelabel when the associated Namespace changes.
+func (r *NamespacelabelReconciler) enqueueRequestsFromNamespace(ctx context.Context, namespace client.Object) []reconcile.Request {
+	ns, ok := namespace.(*corev1.Namespace)
+	if !ok {
+		r.Log.Error(nil, "Failed to cast object to Namespace", "object", namespace)
+		return []reconcile.Request{}
+	}
+
+	namespaceLabelList := &labelsv1.NamespacelabelList{}
+	listOps := &client.ListOptions{
+		Namespace: ns.Name,
+	}
+	if err := r.List(ctx, namespaceLabelList, listOps); err != nil {
+		r.Log.Error(err, "Failed to list Namespacelabel resources", "Namespace", ns.Name)
+		return []reconcile.Request{}
+	}
+
+	var requests []reconcile.Request
+	for _, item := range namespaceLabelList.Items {
+		if item.Namespace == ns.Name {
```

Comment:
You're iterating on a list which you got from the namespace so this will always be true